src/sentry/static/sentry/app/utils/tokenizeSearch.tsx

```diff
@@ -1,93 +1,176 @@
-import flatMap from 'lodash/flatMap';
-
 import {escapeDoubleQuotes} from 'app/utils';
 
-export type QueryResults = {
-  /**
-   * The text portion of the string query
-   */
-  query: string[];
-  [k: string]: string[];
+export enum TokenType {
+  OP,
+  TAG,
+  QUERY,
+}
+
+export type Token = {
+  type: TokenType;
+  key?: string;
+  value: string;
 };
 
-/**
- * Tokenize a search into an object
- *
- * Example:
- *   tokenizeSearch('is:resolved foo bar tag:value');
- *   {
- *     is ['resolved'],
- *     query: ['foo', 'bar'],
- *     tag: ['value'],
- *   }
- *
- * Should stay in sync with src.sentry.search.utils:tokenize_query
- */
-export function tokenizeSearch(query: string) {
-  const tokens = splitSearchIntoTokens(query);
+export class QueryResults {
+  tagValues: object;
+  tokens: Token[];
+
+  constructor(strTokens: string[]) {
+    this.tokens = [];
+    this.tagValues = {};
+    for (let token of strTokens) {
+      let tokenState = TokenType.QUERY;
 
-  const searchParams: {query: string[]; tags: string[]} = {
-    query: [],
-    tags: [],
-  };
+      if (['OR', 'AND'].includes(token.toUpperCase())) {
+        this.addOp(token.toUpperCase());
+        continue;
+      }
+
+      if (token.startsWith('(')) {
+        const parenMatch = token.match(/^\(+/g);
+        if (parenMatch) {
+          this.addOp(parenMatch[0]);
+          token = token.replace(/^\(+/g, '');
+        }
+      }
 
-  for (const token of tokens) {
-    let tokenState: 'query' | 'tags' = 'query';
+      // Traverse the token and determine if it is a tag
+      // condition or bare words.
+      for (let i = 0, len = token.length; i < len; i++) {
+        const char = token[i];
 
-    // Traverse the token and determine if it is a tag
-    // condition or bare words.
-    for (let i = 0, len = token.length; i < len; i++) {
-      const char = token[i];
+        if (i === 0 && (char === '"' || char === ':')) {
+          break;
+        }
 
-      if (i === 0 && (char === '"' || char === ':')) {
-        break;
+        // We may have entered a tag condition
+        if (char === ':') {
+          const nextChar = token[i + 1] || '';
+          if ([':', ' '].includes(nextChar)) {
+            tokenState = TokenType.QUERY;
+          } else {
+            tokenState = TokenType.TAG;
+          }
+          break;
+        }
       }
 
-      // We may have entered a tag condition
-      if (char === ':') {
-        const nextChar = token[i + 1] || '';
-        if ([':', ' '].includes(nextChar)) {
-          tokenState = 'query';
-        } else {
-          tokenState = 'tags';
+      let trailingParen: string = '';
```

Comment:
```suggestion
      let trailingParen = '';
```

Typescript should be able to infer this type.